File: organize_archive/services/dups.py
"""Duplicate groups: the Duplicates panel's summary tile and group listing.

Reads `dup_groups`/`dup_members`, built by `dedup/` -- this module never groups
or picks a canonical itself, only reports what dedup already decided.
"""

from __future__ import annotations

import os
import sqlite3
from typing import Any

from ._common import _root_clause, reading


@reading
def dup_summary(conn: sqlite3.Connection, root_id: int | None = None) -> dict[str, Any]:
    """The Duplicates panel's summary tile: group/duplicate/reclaimable-bytes
    totals, plus the redundant copies broken down by match type (identical vs
    visual) and by media type."""
    rc, rp = _root_clause(root_id)
    row = conn.execute(
        f"""SELECT COUNT(*) groups,
                   COALESCE(SUM(g.member_count-1),0) dups,
                   COALESCE(SUM(g.redundant_bytes),0) bytes
            FROM dup_groups g
            JOIN files f ON f.id=g.canonical_file_id
            WHERE 1=1{rc}""",
        rp,
    ).fetchone()
    # Breakdown of the redundant copies (things_to_fix #35): "24,102
    # duplicates" says nothing about what they are. Two cuts of the same
    # rows, so both add up to `duplicates`/`reclaimable` exactly:
    #
    # * by match -- byte-identical to the kept copy, or only visually the
    #   same (a re-compressed export). Decided per MEMBER, not by the
    #   group's `method`: a perceptual group routinely also contains exact
    #   copies, and this is the same rule the duplicate tiles label
    #   themselves with, so the panel can never contradict them.
    # * by media type -- a hundred redundant videos and a hundred redundant
    #   thumbnails are not the same news.
    #
    # `f` is the canonical here, same as in the count above, so the shared
    # _root_clause keeps filtering on the root the group belongs to; `d` is
    # the redundant copy being described.
    by_match: dict[str, dict[str, int]] = {}
    by_media: dict[str, dict[str, int]] = {}
    for r in conn.execute(
        f"""SELECT CASE WHEN d.sha256 IS NOT NULL AND d.sha256 = f.sha256
                        THEN 'identical' ELSE 'visual' END AS match_type,
                   d.media_type AS media_type,
                   COUNT(*) AS n, COALESCE(SUM(d.size), 0) AS bytes
            FROM dup_members m
            JOIN dup_groups g ON g.id = m.group_id
            JOIN files f ON f.id = g.canonical_file_id
            JOIN files d ON d.id = m.file_id
            WHERE m.role = 'duplicate'{rc}
            GROUP BY match_type, d.media_type""",
        rp,
    ):
        for bucket, key in (
            (by_match, r["match_type"]),
            (by_media, r["media_type"] or "other"),
        ):
            acc = bucket.setdefault(key, {"count": 0, "bytes": 0})
            acc["count"] += r["n"]
            acc["bytes"] += r["bytes"]

    def _ranked(
        bucket: dict[str, dict[str, int]], order: list[str] | None = None
    ) -> list[dict[str, Any]]:
        items: list[dict[str, Any]] = [{"key": k, **v} for k, v in bucket.items()]
        if order:  # fixed order where one exists (identical first)
            items.sort(key=lambda i: order.index(i["key"]) if i["key"] in order else len(order))
        else:
            items.sort(key=lambda i: (-i["count"], i["key"]))
        return items

    return {
        "groups": row["groups"],
        "duplicates": row["dups"],
        "reclaimable": row["bytes"],
        "by_match": _ranked(by_match, ["identical", "visual"]),
        "by_media": _ranked(by_media),
    }


@reading
def dup_groups(
    conn: sqlite3.Connection, root_id: int | None = None, limit: int = 60, offset: int = 0
) -> dict[str, Any]:
    """A page of duplicate groups, largest reclaimable-bytes first, each with
    its canonical and duplicate members (name, folder, role, match type)."""
    rc, rp = _root_clause(root_id)
    groups = conn.execute(
        f"""SELECT g.id, g.method, g.member_count, g.size_each, g.redundant_bytes,
                   g.canonical_file_id
            FROM dup_groups g JOIN files f ON f.id=g.canonical_file_id
            WHERE 1=1{rc}
            ORDER BY g.redundant_bytes DESC, g.id
            LIMIT ? OFFSET ?""",
        (*rp, limit, offset),
    ).fetchall()
    out = []
    for g in groups:
        members = conn.execute(
            """SELECT f.id, f.media_type, f.rel_path, m.role,
                      r.path AS root,
                      CASE
                        WHEN m.role='canonical' THEN 'canonical'
                        WHEN f.sha256 = canonical.sha256 THEN 'identical'
                        ELSE 'visual'
                      END AS match_type
               FROM dup_members m JOIN files f ON f.id=m.file_id
               JOIN roots r ON r.id=f.root_id
               JOIN files canonical ON canonical.id=?
               WHERE m.group_id=? ORDER BY (m.role='duplicate'), f.id""",
            (g["canonical_file_id"], g["id"]),
        ).fetchall()
        out.append(
            {
                "id": g["id"],
                "method": g["method"],
                "count": g["member_count"],
                "size_each": g["size_each"],
                "reclaimable": g["redundant_bytes"],
                "canonical_id": g["canonical_file_id"],
                "members": [
                    {
                        "id": m["id"],
                        "type": m["media_type"],
                        "role": m["role"],
                        "match_type": m["match_type"],
                        "name": os.path.basename(m["rel_path"]),
                        "folder": os.path.dirname(m["rel_path"]),
                    }
                    for m in members
                ],
            }
        )
    return {"groups": out, "offset": offset, "count": len(out)}

File: organize_archive/web/static/js/dups.js
// The Duplicates screen: the summary stats, the breakdown of what the redundant
// copies are, and one row per group.

import {
  jget,
} from "./api.js";
import {
  fmtBytes,
} from "./dom.js";
import {
  S, TYPE_COL, TYPE_ICON, typeLabel,
} from "./state.js";
import {
  startInfiniteList,
} from "./main.js";

const DUP_PAGE_SIZE = 40;
export async function renderDedup(m) {
  const gen = S.nav, root = S.arch.id;
  const ds = await jget("/api/dups/summary?root=" + root);
  if (gen !== S.nav) return;
  if (!ds.groups) {
    m.innerHTML = `<div class="pagehead"><div><h2 class="sec">Duplicates</h2><p>Review redundant copies Archive has safely hidden from your main library.</p></div></div>
      <div class="soonbox"><div class="big">🧹</div>
      <p>No duplicates found yet.</p>
      <p class="muted">This runs automatically once files are scanned and dated. See the Overview tab for progress. Groups byte-identical copies and visually identical image exports (such as re-compressed JPG/PNG/HEIC files); nothing is ever deleted, extra copies are just hidden from browsing.</p></div>`;
    return;
  }
  m.innerHTML = `<div class="pagehead"><div><h2 class="sec">Duplicates</h2><p>Review redundant copies Archive has safely hidden from your main library.</p></div></div>
    <div class="statrow">
      <div class="stat"><div class="k">Duplicate groups</div><div class="v">${ds.groups.toLocaleString()}</div></div>
      <div class="stat"><div class="k">Redundant copies</div><div class="v">${ds.duplicates.toLocaleString()}</div></div>
      <div class="stat"><div class="k">Reclaimable</div><div class="v">${fmtBytes(ds.reclaimable)}</div></div>
    </div>
    ${dupBreakdownPanel(ds)}
    <div class="muted" style="margin-bottom:12px">Exact copies and visually identical image exports. The <span style="color:var(--good)">✓ kept</span> copy stays in Browse; the rest are hidden (never deleted). Biggest space first.</div>
    <div id="dupgroups"></div>
    <div class="infinite-status" id="dup-sentinel" aria-live="polite"></div>`;
  startInfiniteList("dupList", {
    sentinelId: "dup-sentinel", pageSize: DUP_PAGE_SIZE,
    fetchPage: async offset => {
      const res = await jget(`/api/dups?root=${S.arch.id}&offset=${offset}&limit=${DUP_PAGE_SIZE}`);
      return res.groups;
    },
    onPage: (groups, { first }) => {
      const wrap = document.getElementById("dupgroups");
      if (first) wrap.innerHTML = "";
      groups.forEach(g => wrap.appendChild(dupGroupRow(g)));
    },
  });
}
// What the redundant copies actually ARE (things_to_fix #35). "27,318
// duplicates" hides two things worth knowing: how many are byte-identical
// (safe, boring) versus only visually the same (a re-compressed export,
// where the kept copy is a judgement call), and whether the space is going
// to photos or to a much smaller number of videos. Same bar/legend
// vocabulary as the Overview's storage panel, and the media colours are the
// shared TYPE_COL so a hue means the same thing on every screen.
const DUP_MATCH_LABEL = { identical: "Identical copies", visual: "Visual matches" };
const DUP_MATCH_COL = { identical: "#30d158", visual: "#5b8cff" };
function dupBreakdownPanel(ds) {
  const rows = [
    { label: "By match", items: ds.by_match,
      name: k => DUP_MATCH_LABEL[k] || k, colour: k => DUP_MATCH_COL[k] || "#8e8e93" },
    { label: "By media", items: ds.by_media,
      name: k => typeLabel(k), colour: k => TYPE_COL[k] || TYPE_COL.other },
  ].filter(r => (r.items || []).length);
  if (!rows.length) return "";       // older payload, or nothing to break down
  const total = ds.duplicates || 1;
  const section = r => `<div class="type-bar-row">
      <span class="type-bar-label">${r.label}</span>
      <div class="type-summary-bar">${r.items.map(i =>
    `<div class="type-summary-segment" style="width:${100 * i.count / total}%;background:${r.colour(i.key)}" title="${r.name(i.key)}: ${i.count.toLocaleString()} · ${fmtBytes(i.bytes)}"></div>`
  ).join("")}</div></div>
    <div class="type-summary-legend">${r.items.map(i =>
    `<span><span class="type-summary-key" style="background:${r.colour(i.key)}"></span>${r.name(i.key)} <span class="muted">· ${i.count.toLocaleString()} (${(100 * i.count / total).toFixed(1)}%) · ${fmtBytes(i.bytes)}</span></span>`
  ).join("")}</div>`;
  return `<div class="panel dup-breakdown"><div class="panel-heading"><div>
      <h3>What the copies are</h3><p>Every redundant copy, split two ways</p></div></div>
    ${rows.map(section).join("")}</div>`;
}
function dupGroupRow(g) {
  const row = document.createElement("div"); row.className = "dupgroup";
  const head = `<div class="dghead"><b>${g.count}×</b> · ${fmtBytes(g.size_each)} each ·
      <span class="muted">${fmtBytes(g.reclaimable)} reclaimable</span></div>`;
  const tiles = g.members.map(mm => {
    const kept = mm.role === "canonical";
    const tag = kept ? '<span class="keep">✓ kept</span>'
      : mm.match_type === "identical" ? '<span class="duptag exact">Identical copy</span>'
        : '<span class="duptag visual">Visual match</span>';
    const thumb = (mm.type === "image" || mm.type === "video") ? `<img src="/thumb/${mm.id}" loading="lazy" onerror="this.replaceWith(Object.assign(document.createElement('div'),{className:'ph',textContent:'${TYPE_ICON[mm.type] || "📦"}'}))">`
      : `<div class="ph">${TYPE_ICON[mm.type] || "📦"}</div>`;
    return `<div class="duptile ${kept ? 'kept' : ''}" title="${mm.folder}" onclick="openItem(${mm.id})">
        ${thumb}<div class="dtcap">${tag}</div>
        <div class="dtpath">${mm.folder || '/'}</div></div>`;
  }).join("");
  row.innerHTML = head + `<div class="duprow">${tiles}</div>`;
  return row;
}
